tests/components/jewish_calendar/test_binary_sensor.py

```diff
@@ -0,0 +1,107 @@
+        make_nyc_test_params(dt(2018, 9, 9, 21, 25), STATE_ON),
+        make_nyc_test_params(dt(2018, 9, 10, 21, 25), STATE_ON),
+        make_nyc_test_params(dt(2018, 9, 28, 21, 25), STATE_ON),
+        make_nyc_test_params(dt(2018, 9, 29, 21, 25), STATE_OFF),
+        make_nyc_test_params(dt(2018, 9, 30, 21, 25), STATE_ON),
+        make_nyc_test_params(dt(2018, 10, 1, 21, 25), STATE_ON),
+        make_jerusalem_test_params(dt(2018, 9, 29, 21, 25), STATE_OFF),
+        make_jerusalem_test_params(dt(2018, 9, 30, 21, 25), STATE_ON),
+        make_jerusalem_test_params(dt(2018, 10, 1, 21, 25), STATE_OFF),
+    ]
+    melacha_test_ids = [
+        "currently_first_shabbat",
+        "after_first_shabbat",
+        "friday_upcoming_shabbat",
+        "upcoming_rosh_hashana",
+        "currently_rosh_hashana",
+        "second_day_rosh_hashana",
+        "currently_shabbat_chol_hamoed",
+        "upcoming_two_day_yomtov_in_diaspora",
+        "currently_first_day_of_two_day_yomtov_in_diaspora",
+        "currently_second_day_of_two_day_yomtov_in_diaspora",
+        "upcoming_one_day_yom_tov_in_israel",
+        "currently_one_day_yom_tov_in_israel",
+        "after_one_day_yom_tov_in_israel",
+    ]
+
+    @pytest.mark.parametrize(
+        [
+            "now",
+            "candle_lighting",
+            "havdalah",
+            "diaspora",
+            "tzname",
+            "latitude",
+            "longitude",
+            "result",
+        ],
+        melacha_params,
+        ids=melacha_test_ids,
+    )
+    async def test_issur_melacha_sensor(
+        self,
+        hass,
+        now,
+        candle_lighting,
+        havdalah,
+        diaspora,
+        tzname,
+        latitude,
+        longitude,
+        result,
+    ):
+        """Test Issur Melacha sensor output."""
+        time_zone = get_time_zone(tzname)
+        test_time = time_zone.localize(now)
+
+        default_time_zone = str(hass.config.time_zone)
+        hass.config.set_time_zone(tzname)
+        hass.config.latitude = latitude
+        hass.config.longitude = longitude
+
+        assert await async_setup_component(
+            hass,
+            jewish_calendar.DOMAIN,
+            {
+                "jewish_calendar": {
+                    "name": "test",
+                    "language": "english",
+                    "diaspora": diaspora,
+                    "candle_lighting_minutes_before_sunset": candle_lighting,
+                    "havdalah_minutes_after_sunset": havdalah,
+                }
+            },
+        )
+        await hass.async_block_till_done()
+
+        with alter_time(test_time):
+            await hass.helpers.entity_component.async_update_entity(
+                "binary_sensor.test_issur_melacha_in_effect"
+            )
+
+        assert (
+            hass.states.get("binary_sensor.test_issur_melacha_in_effect").state
+            == result
+        )
+        hass.config.set_time_zone(default_time_zone)
```

Comment:
Why do we need to set the time zone here?

Reply:
To restore the time_zone, changing time zones has given me much trouble:

- I started off with having the component reading `hass.config.time_zone`
  - If I try setting the time_zone with `dt_util.set_default_time_zone()` in the test, during async_setup the change is not reflected in `hass.config.time_zone`
  - If I try setting `hass.config.time_zone` using `hass.config.set_time_zone()`, my tests pass, but even though I restore it to the original value, future tests fail (Thinking maybe I needed to await?).

- So then I decided to use in the component `dt_util.DEFAULT_TIME_ZONE` which works with `dt_util.set_default_time_zone()` in the test, but I don't think thats the correct way for a component to get the time_zone?

Reply:
What is the reason we need to change timezone?

Generally, if things need to be setup before testing, it should be done in pytest fixtures. They can support tear down code by using `yield` after things shave been setup.

Reply:
The sensor is time_zone sensitive, I create a hdate.Location object using the external library which takes the time zone info.
I want to check that if change the home assistant timezone, the calculations are done correctly.
I agree about using a fixture, but I need to pass a parameter to the fixture, as I want to use both New York and Jerusalem timezones. I'm not aware of a way to do that.

Reply:
We could have two different fixtures that set different timezones and perhaps also two different tests. The timezone is passed during component setup, correct? So we don't need to check different timezones dynamically during the entity lifetime.

Reply:
Ok, I also finally figured out why the time zones weere not working for me:

The tests.common.hass object sets the time_zone value directly without using the set_time_zone() method. The set_time_zone method has a secondary side effect of setting dt_util.DEFAULT_TIME_ZONE. But the DEFAULT_TIME_ZONE global does not affect hass.config.time_zone.

So when I try changing the time_zone in the setup by changing DEFAULT_TIME_ZONE, it didn't affect the Location object in the Jewish calendar platform.
On the other hand when using the hass.config.set_time_zone() I was affecting the DEFAULT_TIME_ZONE value and thereby breaking later tests. Resetting the value to the previous value of hass.config.time_zone didn't help as the original value of hass.config.time_zone is US\Pacific but the default value of DEFAULT_TIME_ZONE is UTC, so I was resetting DEFAULT_TIME_ZONE to US\Pacific instead of UTC.

I think the correct way in the test is to use hass.config.set_time_zone() and in the tearDown method to reset the DEFAULT_TIME_ZONE to UTC.

Reply:
I think I managed to finally fix the time_zone issue.

I'd appreciate it if you could give the tests another look. Resolving the conversation for now.